### DISCUSSION TITLE
Error in launching pyPrimeMesh

### DISCUSSION THREAD

**Discussion by som1197** [ansys/pyprimemesh#814] [Q&A] [answered]
Hello,

I have my ANSYS license installed on a server (Host) that I am working on remotely using ANSYS 2023 R2 installed on the client. When I try to run the prime client, it gives me the following error.

```
FileNotFoundError                         Traceback (most recent call last)
Cell In[1], line 7
      4 from ansys.meshing import prime
      5 from ansys.meshing.prime.graphics import Graphics
----> 7 prime_client = prime.launch_prime(timeout = 60)
      8 model = prime_client.model
      9 mesh_util = prime.lucid.Mesh(model=model)

File ~\AppData\Local\anaconda3\envs\Ansys_related\Lib\site-packages\ansys\meshing\prime\internals\launcher.py:243, in launch_prime(prime_root, ip, port, timeout, n_procs, version, **kwargs)
    240     client.container_name = container_name
    241     return client
--> 243 server = launch_server_process(
    244     prime_root=prime_root, ip=ip, port=port, n_procs=n_procs, **kwargs
    245 )
    247 return Client(server_process=server, ip=ip, port=port, timeout=timeout)

File ~\AppData\Local\anaconda3\envs\Ansys_related\Lib\site-packages\ansys\meshing\prime\internals\launcher.py:85, in launch_server_process(prime_root, ip, port, n_procs, **kw)
     83     prime_root = get_ansys_prime_server_root()
     84     if prime_root is None:
---> 85         raise FileNotFoundError('No valid Ansys Prime Server found to launch.')
     86 else:  # verify if the file exists
     87     if not os.path.isdir(prime_root):

FileNotFoundError: No valid Ansys Prime Server found to launch.
```
This is the input code that i used.

```
import os
import tempfile

from ansys.meshing import prime
from ansys.meshing.prime.graphics import Graphics

prime_client = prime.launch_prime(timeout = 60)
model = prime_client.model
mesh_util = prime.lucid.Mesh(model=model)
```
This works correctly, when I run this cod snippet using a jupyter notebook installed on my server machine.

Can you let me know if there is a different way to access this? I increased the server timeout to 60 seconds and also used the correct Port/IP to connect to the server. But it did not work.

**Comment by waltersma** [comment]
Hi @som1197,

PyPrimeMesh is the open source client for the Ansys Prime Server that is distributed with the Ansys installation.

If I understand the question correctly you are running the PyPrimeMesh client on a different machine to the Ansys Prime Server instance you wish to use.  Please refer to the docs [here](https://prime.docs.pyansys.com/version/stable/user_guide/launch_prime.html#launch-server-from-a-windows-or-linux-console-and-connect-the-client).  

You will need to launch the Ansys Prime Server from the Y machine and then connect the client to it from X.  The code you are running above is trying to launch the Ansys Prime Server locally and not finding an Ansys installation.  You can check the environment variable AWP_ROOT232 exists and points to a valid installation containing a folder named "meshing" to confirm if the Ansys installation should be available on X.

Once an instance of Ansys Prime Server is started on Y machine (launching from console) you can connect the PyPrimeMesh client on X:

Launch Ansys Prime Server from a console on Y:
```"%AWP_ROOT232%\meshing\Prime\runPrime.bat" server -np 4 --ip 127.0.0.1 --port 50055```

Connect the PyPrimeMesh client on X from the python script:
```
import ansys.meshing.prime as prime

prime_client = prime.Client(ip="127.0.0.1", port=50055)
model = prime_client.model
```

I hope this helps!

**Reply by som1197** [reply]
Hi @waltersma ,

By client I am referring to workstation (X) and the host machine is the server (Y) on which all the ANSYS license files have been installed.  Currently the ANSYS license manager in X is configured to simply point to the main server Y. I am able to use the Prime client on the sever Y independent of the ip or port. But if I do this from the workstation X, it gives me the error shown previously.

So, I tried using your recommendation and opened up a prime client on the server Y. But I want to use pyPrimeMesh from the workstation X. Now, when I try to call the prime client there, it doesn't work.

Does this provide more clarity regarding the issue?

Thanks!

**Comment by waltersma** [comment]
Hi @som1197,

I edited my response to hopefully be clearer.  The original error you are seeing is that launching locally the Ansys Prime Server it cannot find the Ansys 23R2 installation.  

When you say you tried the recommendation and it didn't work, what was the error?  Was it to do with a timeout?  If so, please check that you can ping the machine OK and that the Ansys Prime Server launched ok and did not close for any reason.

Thanks

**Reply by som1197** [reply]
Hi @waltersma ,

There are 2 approaches that I used.
1) Using this shell script via windows cmd to start the prime server. "C:\Program Files\ANSYS Inc\v241\meshing\Prime\runPrime.bat" server -np 4 --ip 10.0.0.41 --port 50055 - This allowed me to use 4 threads and started the ANSYSPrimeServer. Next, I used the following code on the client side to connect with it remotely.

```
import os
import tempfile

from ansys.meshing import prime
from ansys.meshing.prime.graphics import Graphics

prime_client = prime.Client(ip = "10.0.0.41", port=50055, timeout = 60)
model = prime_client.model

mesh_util = prime.lucid.Mesh(model=model)
file_name = r'D:\08_Mazzei_internal_projects\16_Parametric_meshes\01_pyPrimeMesh\03_Geom\01_CMR\case_3a.1.scdoc'
params = prime.ImportCadParams(
    model=model, cad_reader_route=prime.CadReaderRoute.SPACECLAIM
)
prime.FileIO(model).import_cad(file_name=file_name, params=params)
```

This allows me to run the prime_client but I am unable to read any .scdoc files. I opened a separate discussion for that.

2) In the second approach I replaced the shell script on the server side by this code,

```
import os
import tempfile

from ansys.meshing import prime
from ansys.meshing.prime.graphics import Graphics

prime_client = prime.launch_prime(n_procs=4, timeout=60)
model = prime_client.model
```

But when I try to connect the prime server from the client side, it doesn't work.

Let me know if you need any other clarifications. Also I am running a 2023 R2 version on the client side and 2024 R1 version on the server side.

**Reply by som1197** [reply, marked as answer]
Hi @waltersma ,

I was able to resolve this issue. Thanks again for the help.

**Reply by som1197** [reply]
A few points that helped in resolving this issue are listed below.

I was using the same code snippet to run the prime client on the client and he server side. The correct way of doing this is:

- Use this code to start the prime server (If an actual server is being used)

`"%AWP_ROOT242%\meshing\Prime\runPrime.bat" server -np 4 --ip 127.0.0.1 --port 50055` OR
```
import ansys.meshing.prime as prime
prime_client = prime.launch_prime()
```
- Use this code to start the prime client and connect to the server (From a remote workstation). Use the ip address of the server location.
```
import ansys.meshing.prime as prime
prime_client = prime.Client(ip="127.0.0.1", port=50055)
model = prime_client.model
```
Since the traffic on the port 50055 was being blocked, I had to allow the app to send and receive data on Windows firewall. I followed these steps to do so.
- Open Windows Defender Firewall and on the sidebar you'll see a message "Allow and app or feature through Windows Defender Firewall", then select "Change settings"
- Inside the allowed apps, you should see the AnsysPrimeServer already listed. Select the type of network the devices are on "Domain, Private, etc." (You can verify this in network settings)
- Save those changes and try again. (You may need to restart) 
- If the ansysprimeserver isn't there, in Task manager, right click the process, "AnsysPrimeServer", and select "Open File Location", take note of the path where AnsysPrimeServer.exe is located.
- Towards the bottom, select "Allow another app...", then select "Browse" and navigate to the directory where AnsysPrimeServer.exe is located. Select it and click open, then add the app.
- Then add the app. Select the type of network the devices are on "Domain, Private, etc." (You can verify this in network settings)
- Save those changes and try again. (You may need to restart)
